### PR TITLE
[red-knot] Simplify tuples containing `Never`

### PR DESCRIPTION
**Pull request by sharkdp**
## Summary

Simplify tuples containing `Never` to `Never`:

```py
from typing import Never

def never() -> Never: ...

reveal_type((1, never(), "foo"))  # revealed: Never
```

I should note that mypy and pyright do *not* perform this simplification. I don't know why.

There is [only one place](https://github.com/astral-sh/ruff/blob/5137fcc9c81610f687b6cb45413ef83c2c5eea73/crates/red_knot_python_semantic/src/types/infer.rs#L1477-L1484) where we use `TupleType::new` directly (instead of `Type::tuple`, which changes behavior here). This appears when creating `TypeVar` constraints, and it looks to me like it should stay this way, because we're using `TupleType` to store a list of constraints there, instead of an actual type. We also store `tuple[constraint1, constraint2, …]` as the type for the `constraint1, constraint2, …` tuple expression. This would mean that we infer a type of `tuple[str, Never]` for the following type variable constraints, without simplifying it to `Never`. This seems like a weird edge case that's maybe not worth looking further into?!
```py
from typing import Never

#         vvvvvvvvvv
def f[T: (str, Never)](x: T):
    pass
```

## Test Plan

- Added a new unit test. Did not add additional Markdown tests as that seems superfluous.
- Tested the example above using red knot, mypy, pyright.
- Verified that this allows us to remove `contains_never` from the property tests (https://github.com/astral-sh/ruff/pull/14178#discussion_r1866473192)

### CODE REVIEW

**Review by carljm — APPROVED** [review]
Looks great!

File: crates/red_knot_python_semantic/src/types.rs [hunk 1]
```diff
@@ -574,8 +574,11 @@ impl<'db> Type<'db> {
...
     }
 
-    pub fn tuple(db: &'db dyn Db, elements: &[Type<'db>]) -> Self {
-        Self::Tuple(TupleType::new(db, elements))
+    pub fn tuple<T: Into<Type<'db>>>(
+        db: &'db dyn Db,
+        elements: impl IntoIterator<Item = T>,
+    ) -> Self {
+        TupleType::from_elements(db, elements)
     }
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:582]: nit: this method now feels a bit redundant. I'd probably have got rid of it and changed the callsites to use the new `TupleType::from_elements()` method

**Comment by AlexWaygood** [comment]
> I should note that mypy and pyright do _not_ perform this simplification. I don't know why.

Mypy in general appears not to do much eager simplification at all, e.g. it doesn't even simplify this union:

```py
def f(x: int | int):
    reveal_type(x)  # revealed type is "Union[builtins.int, builtins.int]"
```

Pyright does a lot more eager simplification than mypy. But it's important to note, I think, that it's only relatively recently that we formalized the idea that `Never`/`NoReturn` constitute the bottom type in Python. Mypy has historically always treated it as such, but I think pyright only started treating it as such [relatively recently](https://github.com/microsoft/pyright/issues/2921). So the issue of simplification may just not have been considered. This is just speculation, though!
